README.md
This is master branch of testdroid-documentation. You are probably looking for [Testdroid documentation](http://bitbar.github.io/testdroid-documentation).


testdroid-cloud-ui/release_notes.md
---
layout: default
title: Releas Notes
---

# CHANGES IN VERSION 2.8 

## General

* New activation wizard for cloud users  
* Interactive supports now rotating device.  
* Support for generating API key for authentication in API test
  runs. No need to share user credentials anymore.  
* Support for FPS for Appium tests.  
* And a whole lot of smaller improvements and fixes.

# CHANGES IN VERSION 2.7

## General

* Enable more FREE devices for cloud users.  

## Subscriptions

* Improve billing and usage information visualization  
* Shortcut to Buy and Upgrade plans.  
* Improve upgrade plan process  

## Test Runs

* Filtering enhancements  
* Add device specific retry  

## Appium

* Start tests immediately after script is launched  
* More descriptive error messages  


# CHANGES IN VERSION 2.6 

## General

* Modal improvements
* Changes in header and footer
* Change email notifications

## Interactive

* Simplify start procedure
* Add search to logcat -view
* Pause/start to logcat -view
* Show plain text on logcat- -view

## Screenshots

* Improvements for taking screenshots

## Private Cloud

* Enable new deployment options
* Fully customizable UI
* Improved audit logging
* Add support uploading huge apk/ipa files


# CHANGES IN VERSION 2.5 

## Appium

* Version update
* Chromedriver update

## Interactive

* Usability improvements

## Test Run -view

* Filtering improvements
* Layout changes

## User Management

* Flexible role based user management
* Enable APPIUM & IOS for all users

## Subscription

* Usage details
* Billing details
* Improved pricing information 


# CHANGES IN VERSION 2.4 

## General

* Feedback form
* Add shortcut to tutorial videos in help

## My Account

* Select default -view

## Device Groups

* Show public groups

## Test Run -view

* Test case -comparison
* Search filter
* Hide excluded devices by default

## Device Run -view

* FPS performance data

## Android

* Appium -parallel run for Android


# CHANGES IN VERSION 2.3

## General

* Add menu, if display width > 1024 pxl
* Background and login changed to white
* Set projects-view as the default
* Show the latest project test run by default
* Changed first time wizard.
* (?) Help pages renewal linkgs help page

## Test Run -view

* Summary widget

## Device Run view

* Shortcut to change device
* Stack trace

## Interactive

* Logcat -logs
* Full screen -mode

## RiC

* Improved test results coverage


js/toc.js
// https://github.com/ghiculescu/jekyll-table-of-contents
(function($){
  $.fn.toc = function(options) {
    var defaults = {
      noBackToTopLinks: false,
      title: '<h1><i>Contents</i></h1>',
      minimumHeaders: 3,
      headers: 'h1, h2, h3, h4, h5, h6',
      listType: 'ul', // values: [ol|ul]
      showEffect: 'none', // values: [show|slideDown|fadeIn|none]
      showSpeed: 'slow' // set to 0 to deactivate effect
    },
    settings = $.extend(defaults, options);

    function fixedEncodeURIComponent (str) {
      return encodeURIComponent(str).replace(/[!'()*]/g, function(c) {
        return '%' + c.charCodeAt(0).toString(16);
      });
    }

    var headers = $(settings.headers).filter(function() {
      // get all headers with an ID
      var previousSiblingName = $(this).prev().attr( "name" );
      if (!this.id && previousSiblingName) {
        this.id = $(this).attr( "id", previousSiblingName.replace(/\./g, "-") );
      }
      return this.id;
    }), output = $(this);
    if (!headers.length || headers.length < settings.minimumHeaders || !output.length) {
      $(this).hide();
      return;
    }

    if (0 === settings.showSpeed) {
      settings.showEffect = 'none';
    }

    var render = {
      show: function() { output.hide().html(html).show(settings.showSpeed); },
      slideDown: function() { output.hide().html(html).slideDown(settings.showSpeed); },
      fadeIn: function() { output.hide().html(html).fadeIn(settings.showSpeed); },
      none: function() { output.html(html); }
    };

    var get_level = function(ele) { return parseInt(ele.nodeName.replace("H", ""), 10); }
    var highest_level = headers.map(function(_, ele) { return get_level(ele); }).get().sort()[0];
    var return_to_top = '<i class="icon-arrow-up back-to-top"> </i>';

    var level = get_level(headers[0]),
      this_level,
      html = settings.title + " <"+settings.listType+">";
    headers.on('click', function() {
      if (!settings.noBackToTopLinks) {
        window.location.hash = this.id;
      }
    })
    .addClass('clickable-header')
    .each(function(_, header) {
      this_level = get_level(header);
      if (!settings.noBackToTopLinks && this_level === highest_level) {
        $(header).addClass('top-level-header').after(return_to_top);
      }
      if (this_level === level) // same level as before; same indenting
        html += "<li><a href='#" + fixedEncodeURIComponent(header.id) + "'>" + header.innerHTML + "</a>";
      else if (this_level <= level){ // higher level than before; end parent ol
        for(i = this_level; i < level; i++) {
          html += "</li></"+settings.listType+">"
        }
        html += "<li><a href='#" + fixedEncodeURIComponent(header.id) + "'>" + header.innerHTML + "</a>";
      }
      else if (this_level > level) { // lower level than before; expand the previous to contain a ol
        for(i = this_level; i > level; i--) {
          html += "<"+settings.listType+"><li>"
        }
        html += "<a href='#" + fixedEncodeURIComponent(header.id) + "'>" + header.innerHTML + "</a>";
      }
      level = this_level; // update for the next one
    });
    html += "</"+settings.listType+">";
    if (!settings.noBackToTopLinks) {
      $(document).on('click', '.back-to-top', function() {
        $(window).scrollTop(0);
        window.location.hash = '';
      });
    }

    render[settings.showEffect]();
  };
})(jQuery);


js/main.js
baseURL = location.protocol + "//" + location.host + "/testdroid-documentation";
//Create a module using an IIFE
/* Search code novelty of Josh Beam */

;(function(global,$) {
  /*
    Put ourselves into "strict" mode
    This just helps us write cleaner JavaScript
  */
  'use strict';

  Query.prototype = {
    // this.q is our search query (for example, "javascript tutorial")
    set: function(val) {
      this.q = val;
      return this;
    },
    // brings us to our search page with a query string attached
    goToLocation: function(route) {
      if(typeof this.q !== 'undefined' && typeof this.q === 'string') {
        document.location.href=route+'?query='+this.q;
      } else {
        return;
      }
    },
    // returns our search query (for example, "javascript tutorial")
    get: function() {
      return this.q;
    },
    // "grab" the query from the query string in the URL and set this.q to it
    setFromURL: function(name) {
      name = name.replace(/[\[]/, "\\[").replace(/[\]]/, "\\]");
      var regex = new RegExp("[\\?&]" + name + "=([^&#]*)"),
          results = regex.exec(location.search);

      this.q = results === null ? "" : decodeURIComponent(results[1].replace(/\+/g, " "));

      return this;        
    },
    // a wrapper for jQuery's $.get
    getJSON: function(file) {
      return $.get(file);
    }
  };

  // when we initialize a query, we have the option of giving it a query string
  function Query(q) {
    if(typeof q !== 'undefined' && typeof q === 'string') {
      this.q = q;
    }
  }

  // attach the Query object to the window
  global.Query = Query;
})(this,$);

$(document).ready(function(){
    // search.js
    $(function(Query) {
            'use strict';

            var query = new Query();

            $('.search').on('submit', function(e) {
                    // stop the form from doing its default behavior
                    e.preventDefault();

                    // set the query, and go to the search page with our query URL
                    query
                    .set($('.search-box').val().trim())
                    .goToLocation(baseURL + '/search');
            });
    }(Query));
    
    /* End of search related code */
    
    /* Open navigation element if sub element is active, also open sub navigation if exists */
    var activeNavigationElement = $("section .navigation .navigation-closed .active");
    if (activeNavigationElement.length == 1){
        var parents = $(activeNavigationElement).parents(".navigation-closed");
        parents.removeClass("navigation-closed").addClass("navigation-open");
    }
    
    $("section .navigation .navigation-icon").click(function(){
        var parent = $(this).parent("li");
        if (parent.hasClass("navigation-closed")){
            parent.removeClass("navigation-closed").addClass("navigation-open");
        }
        else if (parent.hasClass("navigation-open")){
            parent.addClass("navigation-closed").removeClass("navigation-open");
        }
    }); 
    
    $( "#resizable-navigation" ).resizable({
        handles: 'e',
        maxWidth: 350
    });
    
    // hide resize icon if not touch device
    if (!Modernizr.touch){
        $(".navigation-control-placeholder .navigation-control").hide();
    }
});


js/search-worker.js
//Modification of http://frontendcollisionblog.com/javascript/jekyll/tutorial/2015/03/26/getting-started-with-a-search-engine-for-your-site-no-server-required.html by Josh Beam
var isWorker = true;
    
onmessage = function(e) {
	importScripts('lunr.min.js');

    /* New query returs object data, read from memory returns results */
    if (e.data.data){
		processQueryInWorker(e.data);
	}
	else{
		processResultsInWorker(e.data);
	}
};

function launchOutsideWorker(vars){	
	isWorker = false;

	/* New query returs object data, read from memory returns results */
    if (vars.data){
		processQueryInWorker(vars);
	}
	else{
		processResultsInWorker(vars);
	}
}
    

function processQueryInWorker(vars){
    var data = vars.data,
        query = vars.query,
        site = vars.site,
        contentMaxLength = 400, // Max lenght of content snippet
        pageParam = vars.pageParam,
		queryParam = vars.queryParam,
		queryWords = vars.queryWords,
		searchIndex,
		results,
		totalScore = 0,
		percentOfTotal;

    // set up the allowable fields
    searchIndex = lunr(function() {
            this.field('title');
            this.field('content');
            this.ref('url');
    });
  
    for(var i = 0; i < data.length; i++){
        searchIndex.add(data[i]); 
    }

    // search for the query and store the results as an array
    results = searchIndex.search(query.q);

    // add the title of each post into each result, too (this doesn't come standard with lunr.js)
    for(var result in results) {                      
        for(var dataIndex in data){
            if (results[result].ref === data[dataIndex].url){
                results[result].title = data[dataIndex].title;
                results[result].content = data[dataIndex].content.substr(0, contentMaxLength);
            }
        }
    } 

    for(var i = 0; i < results.length; i++){
        totalScore+=results[i].score;
    }  

    var processData = {pageParam: pageParam, results: results, 
		queryWords: queryWords, queryParam: queryParam};
    
    processResultsInWorker(processData);
    
    // Save results to local storage or cookie, so no need to refetch and process
    if (isWorker){
		postMessage({query: decodeURI(queryParam), results: results});
	}
	else{
		notWorkerPostMessage({query: decodeURI(queryParam), results: results});
	}
}

function processResultsInWorker(data){

    var pageParam = data.pageParam,
		results = data.results,
		queryWords = data.queryWords,
		queryParam = data.queryParam,
		startingIndex = 1,
		visibleResults = 10; // Results per page
    
    if (pageParam && !isNaN(pageParam)){                    
        startingIndex = pageParam;
    }

    startingIndex = (startingIndex - 1) * visibleResults;
    var endingIndex = startingIndex + visibleResults;

    var processKeywords = function(contentString){
        // Thanks to Ryan O'Hara http://stackoverflow.com/questions/17264639/replace-text-but-keep-case
        function matchCase(text, pattern) {
            var result = '';

            for(var i = 0; i < text.length; i++) {
                var c = text.charAt(i);
                var p = pattern.charCodeAt(i);

                if(p >= 65 && p < 65 + 26) {
                    result += c.toUpperCase();
                } else {
                    result += c.toLowerCase();
                }
            }

            return result;
        }        

        for (var index in queryWords){
            r = new RegExp( "(" + queryWords[index] + ")" , 'gi' );
            contentString = contentString.replace(r, function(match) {
                return "<b>"+matchCase(queryWords[index], match)+"</b>";
            });
        }

        return contentString;
    };
    
    var searchResults = [];
    for (var i = startingIndex; i <= endingIndex; i++){
        var result = results[i];

        if (i >= results.length){
            break;
        }

        if (i < endingIndex){

            //Process keywords
            if (queryWords.length){
                result.content = processKeywords(result.content);
            }
            
            searchResults.push(result);
        }
    }
    var amountOfPages = Math.ceil(results.length / visibleResults);
    searchResults.push(amountOfPages);
    searchResults.push(queryParam);
    
    postableData = {displayableResults: searchResults};
    if (isWorker){
		postMessage(postableData);
	}
	else{
		notWorkerPostMessage(postableData);
	}
}


js/search-results.js
/* Script logic
 * 1. Do search in a worker not to block script main thread
 * 2. If worker isn't available fallback to just run the functions outside worker, will block main thread
 * 3. Save results into sessionStorage to lessen the re-prosessing
 * 4. If sessionStorage isn't available or is empty process everything between each load (heavy)
 * */

//Modification of http://frontendcollisionblog.com/javascript/jekyll/tutorial/2015/03/26/getting-started-with-a-search-engine-for-your-site-no-server-required.html by Josh Beam    
var globalVariables = {
	pageParam: "",
	queryParam: "",
	site: "",
	queryWords: ""
}    

$(document).ready(function(){
	
	setGlobalVariables();

    $(function(Query) {
        var query = new Query();
		   
		query.setFromURL('query');
		var vars = {query: query, site: globalVariables.site, 
				pageParam: globalVariables.pageParam, 
				queryWords: globalVariables.queryWords, 
				queryParam: globalVariables.queryParam};
		
		/* Check if specific query has been made previously */
		var readResult = readSearchResults(query.q);
		
		if (!readResult){
			/* NOTICE: getJSON url needs to be fixed */   
			query.getJSON(baseURL + '/search.json').done(function(data) {
				vars.data = data;
				handleResults(vars);
			});
		}
		else {
			$.extend(vars, {results: readResult});			
			handleResults(vars);
		}		
    }(Query));
});

function handleResults(vars){
	/* Use search-worker.js also without worker available, use just as a normal script */
	
	if(typeof(Worker) !== "undefined") {
		// call worker function from search index, easier to have it there to modify script url by liquid
		useSearchWorker();
		w.postMessage(vars);
	}
	else{
		workerFallback(vars);
	}
}

/* Handlers for communication with search-worker script
 * search/index.md functions as an intermediate because of useful liquid 
 * funtionalities of getting access to base url */

function notWorkerPostMessage(data){
	handleWorkerMessage(data);
}

function handleWorkerMessage(data){    
    if (data.displayableResults){		
		displayWorkerResults(data.displayableResults); 
	}
	else if (data.query && data.results){
		writeSearchResults(data);
	}
}

/* DOM functionalities with existing results as Web Workers have no
 * access to DOM */
 
function displayWorkerResults(data){
	function decodeEntities(encodedString) {
        var textArea = document.createElement('textarea');
        textArea.innerHTML = encodedString;
        return textArea.value;
    }
    
    var emptyResult = true;
    
    var $results = $('.search-results');

    for(var i = 0; i <= data.length; i++){
        var result = data[i];
        if (typeof result == "object"){
            emptyResult = false;
            result.content = decodeEntities(result.content);
            $results.append('<li class="search-result"><a href="'+ globalVariables.site + result.ref +'">'+result.title+'</a><p>'+decodeURI(result.content)+'</p></li>');
        }
        else{
            if (typeof result == "number" && result > 1){       
                pages = function(){
                    element = "";
                    for (var j = 1; j <= result; j++){
                        if (element.length){
                            element += " / ";
                        }
                        if (j == globalVariables.pageParam || (j == 1 && typeof globalVariables.pageParam === "undefined") ){
                            element += "<span>"+j+"</span>";
                        }
                        else{
                            element += "<a href='"+baseURL+"/search?query="+globalVariables.queryParam+"&page="+j+"'>"+j+"</a>";
                        }
                    }
                    return element;
                }

                $results.append(pages());
            }
            else if (typeof result == "string"){
                $(".content .title").append(' for "'+decodeURI(globalVariables.queryParam)+'"');
                if (emptyResult){
                    $results.append('<li class="search-result"><p>No results for "'+decodeURI(globalVariables.queryParam)+'".</p></li>');
                }
            }
        }   
    }  
}

/* Helper functions */

function setGlobalVariables(){
	var params = window.location.search.substring(1);
    var vars = params.split("&");

    var pageParam = 1, 
		queryParam = "",
		queryWords = [];
		
    for (var k=0;k<vars.length;k++) {
        var helper = vars[k].split("=");
        if (helper[0] == "page"){
            pageParam = parseInt(helper[1]);
        }
        if (helper[0] == "query"){
            queryParam = helper[1];
            queryWords = decodeURI(helper[1]).split(" ");
        }
    }
    
    var site = location.protocol + "//" + location.host;
    
    globalVariables = {
		pageParam: pageParam,
		queryParam: queryParam,
		site: site,
		queryWords: queryWords
	} 
}

/* Write results functionalities:
 * Use session storage as it has been default functionality of browsers for long.
 * No point to see the extra effort to create a cookie fallback */
 
function writeSearchResults(data){
	if (typeof(Storage) !== "undefined"){
		sessionStorage.setItem(data.query, JSON.stringify(data.results));
	}
}

/* Read results functionalities:
 * Use session storage as it has been default functionality of browsers for long.
 * No point to see the extra effort to create a cookie fallback */

function readSearchResults(query){
	if (typeof(Storage) !== "undefined"){
		var data = JSON.parse(sessionStorage.getItem(query));
		
		if (data != null){
			return data;
		}
	}
	return false;
}

/* Function to import search-worker script incase Web Worker is not supported */

// http://stackoverflow.com/questions/8586446/dynamically-load-external-javascript-file-and-wait-for-it-to-load-without-usi thanks to Jason Sebring
function getScript(src, callback) {
	var s = document.createElement('script');
	s.src = src;
	s.async = true;
	s.onreadystatechange = s.onload = function() {
		if (!callback.done && (!s.readyState || /loaded|complete/.test(s.readyState))) {
			callback.done = true;
			callback();
		}
	};
	document.querySelector('head').appendChild(s);
}
